homeassistant/components/synology_dsm/coordinator.py

```diff
@@ -0,0 +1,148 @@
+
+    async def _async_update_data(self) -> dict[str, dict[str, SynoCamera]] | None:
+        """Fetch all data from api."""
+        surveillance_station = self.api.surveillance_station
+        return {
+            "switches": {
+                "home_mode": await self.hass.async_add_executor_job(
+                    surveillance_station.get_home_mode_status
+                )
+            }
+        }
+
+
+class SynologyDSMCentralUpdateCoordinator(SynologyDSMUpdateCoordinator):
+    """DataUpdateCoordinator to gather data for a synology_dsm central device."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        api: SynoApi,
+    ) -> None:
+        """Initialize DataUpdateCoordinator for central device."""
+        super().__init__(
+            hass,
+            entry,
+            api,
+            timedelta(
+                minutes=entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+            ),
+        )
+
+    async def _async_update_data(self) -> dict[str, dict[str, SynoCamera]] | None:
+        """Fetch all data from api."""
+        try:
+            await self.api.async_update()
+        except SYNOLOGY_CONNECTION_EXCEPTIONS as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+        return None
+
+
+class SynologyDSMCameraUpdateCoordinator(SynologyDSMUpdateCoordinator):
+    """DataUpdateCoordinator to gather data for a synology_dsm cameras."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        api: SynoApi,
+    ) -> None:
+        """Initialize DataUpdateCoordinator for cameras."""
+        super().__init__(hass, entry, api, timedelta(seconds=30))
+
+    async def _async_update_data(self) -> dict[str, dict[str, SynoCamera]] | None:
+        """Fetch all camera data from api."""
+        surveillance_station = self.api.surveillance_station
+        current_data: dict[str, SynoCamera] = {
+            camera.id: camera for camera in surveillance_station.get_all_cameras()
+        }
+
+        try:
+            async with async_timeout.timeout(30):
+                await self.hass.async_add_executor_job(surveillance_station.update)
+        except SynologyDSMAPIErrorException as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+
+        new_data: dict[str, SynoCamera] = {
+            camera.id: camera for camera in surveillance_station.get_all_cameras()
+        }
+
+        for cam_id, cam_data_new in new_data.items():
+            if (
+                (cam_data_current := current_data.get(cam_id)) is not None
+                and cam_data_current.live_view.rtsp != cam_data_new.live_view.rtsp
+            ):
+                async_dispatcher_send(
+                    self.hass,
+                    f"{SIGNAL_CAMERA_SOURCE_CHANGED}_{self.entry.entry_id}_{cam_id}",
+                    cam_data_new.live_view.rtsp,
+                )
```

Comment:
It might be better in the future to refactor the camera to a coordinator entity and check for its source changing and fire instead of using a signal, but I didn't change it here.